File: app/src/main/java/com/classicsviewer/app/export/TextExporter.kt
package com.classicsviewer.app.export

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Typeface
import android.graphics.pdf.PdfDocument
import com.classicsviewer.app.models.TextLine
import com.classicsviewer.app.models.TranslationSegment
import java.io.OutputStream

enum class ExportFormat { PDF, TXT, CSV }
enum class ContentType { SOURCE, TRANSLATION }

data class ExportRequest(
    val format: ExportFormat,
    val contentType: ContentType,
    val authorName: String,
    val workTitle: String,
    val bookLabel: String,
    val workId: String,
    val startLine: Int,
    val endLine: Int,
    val translator: String? = null,
    val language: String = "greek"
)

class TextExporter(private val context: Context) {

    fun exportToTxt(
        outputStream: OutputStream,
        request: ExportRequest,
        lines: List<TextLine>? = null,
        translations: List<TranslationSegment>? = null
    ) {
        outputStream.bufferedWriter(Charsets.UTF_8).use { writer ->
            // Write header
            writer.appendLine("${request.authorName} - ${request.workTitle}")
            writer.appendLine("${request.bookLabel}, Lines ${request.startLine}-${request.endLine}")
            if (request.translator != null) {
                writer.appendLine("Translation by ${request.translator}")
            }
            writer.appendLine("Exported from ${request.workId}")
            writer.appendLine()

            // Write content based on type
            when (request.contentType) {
                ContentType.SOURCE -> writeSourceLines(writer, lines ?: emptyList())
                ContentType.TRANSLATION -> writeTranslation(writer, translations ?: emptyList())
            }
        }
    }

    private fun writeSourceLines(writer: java.io.BufferedWriter, lines: List<TextLine>) {
        for (line in lines) {
            val lineNum = line.lineNumber.toString().padStart(5)
            val speaker = if (!line.speaker.isNullOrEmpty()) "[${line.speaker}] " else ""
            writer.appendLine("$lineNum  $speaker${line.text}")
        }
    }

    private fun writeTranslation(writer: java.io.BufferedWriter, segments: List<TranslationSegment>) {
        for (segment in segments) {
            val range = if (segment.endLine != null && segment.endLine != segment.startLine) {
                "[${segment.startLine}-${segment.endLine}]"
            } else {
                "[${segment.startLine}]"
            }
            val speaker = if (!segment.speaker.isNullOrEmpty()) "[${segment.speaker}] " else ""
            writer.appendLine("$range $speaker${segment.translationText}")
            writer.appendLine()
        }
    }

    fun exportToCsv(
        outputStream: OutputStream,
        request: ExportRequest,
        lines: List<TextLine>? = null,
        translations: List<TranslationSegment>? = null
    ) {
        outputStream.bufferedWriter(Charsets.UTF_8).use { writer ->
            // CSV header
            writer.appendLine("line_number,text")

            // Write content based on type
            when (request.contentType) {
                ContentType.SOURCE -> {
                    lines?.forEach { line ->
                        val text = csvEscape(line.text)
                        writer.appendLine("${line.lineNumber},$text")
                    }
                }
                ContentType.TRANSLATION -> {
                    translations?.forEach { segment ->
                        val lineNum = if (segment.endLine != null && segment.endLine != segment.startLine) {
                            "${segment.startLine}-${segment.endLine}"
                        } else {
                            "${segment.startLine}"
                        }
                        val text = csvEscape(segment.translationText)
                        writer.appendLine("$lineNum,$text")
                    }
                }
            }
        }
    }

    private fun csvEscape(text: String): String {
        // If text contains comma, quote, or newline, wrap in quotes and escape quotes
        return if (text.contains(",") || text.contains("\"") || text.contains("\n")) {
            val escaped = text.replace("\"", "\"\"")
            "\"$escaped\""
        } else {
            text
        }
    }

    fun exportToPdf(
        outputStream: OutputStream,
        request: ExportRequest,
        lines: List<TextLine>? = null,
        translations: List<TranslationSegment>? = null
    ) {
        val generator = PdfGenerator(context)
        generator.generate(outputStream, request, lines, translations)
    }
}

class PdfGenerator(private val context: Context) {

    companion object {
        private const val PAGE_WIDTH = 595   // A4 width in points
        private const val PAGE_HEIGHT = 842  // A4 height in points
        private const val MARGIN = 50f
        private const val LINE_HEIGHT = 18f
        private const val TITLE_SIZE = 18f
        private const val SUBTITLE_SIZE = 12f
        private const val BODY_SIZE = 11f
    }

    private val defaultFont: Typeface = Typeface.create("serif", Typeface.NORMAL)
    private val boldFont: Typeface = Typeface.create("serif", Typeface.BOLD)

    fun generate(
        outputStream: OutputStream,
        request: ExportRequest,
        lines: List<TextLine>?,
        translations: List<TranslationSegment>?
    ) {
        val document = PdfDocument()
        var pageNumber = 1
        var yPosition = MARGIN
        var currentPage: PdfDocument.Page? = null
        var canvas: Canvas? = null

        val titlePaint = Paint().apply {
            typeface = boldFont
            textSize = TITLE_SIZE
            isAntiAlias = true
            color = 0xFF000000.toInt()
        }

        val subtitlePaint = Paint().apply {
            typeface = defaultFont
            textSize = SUBTITLE_SIZE
            isAntiAlias = true
            color = 0xFF444444.toInt()
        }

        val bodyPaint = Paint().apply {
            typeface = defaultFont
            textSize = BODY_SIZE
            isAntiAlias = true
            color = 0xFF000000.toInt()
        }

        val speakerPaint = Paint().apply {
            typeface = Typeface.create("serif", Typeface.ITALIC)
            textSize = BODY_SIZE
            isAntiAlias = true
            color = 0xFF666666.toInt()
        }

        fun startNewPage(): Canvas {
            currentPage?.let { document.finishPage(it) }
            val pageInfo = PdfDocument.PageInfo.Builder(PAGE_WIDTH, PAGE_HEIGHT, pageNumber++).create()
            currentPage = document.startPage(pageInfo)
            yPosition = MARGIN
            return currentPage!!.canvas
        }

        fun ensureSpace(needed: Float): Canvas {
            if (yPosition + needed > PAGE_HEIGHT - MARGIN) {
                return startNewPage()
            }
            return canvas!!
        }

        canvas = startNewPage()

        // Draw title
        canvas.drawText("${request.authorName} - ${request.workTitle}", MARGIN, yPosition + TITLE_SIZE, titlePaint)
        yPosition += TITLE_SIZE + 8

        // Draw subtitle
        var subtitle = "${request.bookLabel}, Lines ${request.startLine}-${request.endLine}"
        canvas.drawText(subtitle, MARGIN, yPosition + SUBTITLE_SIZE, subtitlePaint)
        yPosition += SUBTITLE_SIZE + 4

        if (request.translator != null) {
            canvas.drawText("Translation by ${request.translator}", MARGIN, yPosition + SUBTITLE_SIZE, subtitlePaint)
            yPosition += SUBTITLE_SIZE + 4
        }

        canvas.drawText("Exported from ${request.workId}", MARGIN, yPosition + SUBTITLE_SIZE, subtitlePaint)
        yPosition += SUBTITLE_SIZE + 24  // Space before content

        // Draw content
        when (request.contentType) {
            ContentType.SOURCE -> {
                lines?.forEach { line ->
                    canvas = ensureSpace(LINE_HEIGHT)
                    val lineNum = line.lineNumber.toString().padStart(5)

                    // Draw line number
                    canvas!!.drawText(lineNum, MARGIN, yPosition + BODY_SIZE, bodyPaint)

                    // Draw speaker if present
                    var xOffset = MARGIN + 45f
                    if (!line.speaker.isNullOrEmpty()) {
                        val speakerText = "[${line.speaker}] "
                        canvas!!.drawText(speakerText, xOffset, yPosition + BODY_SIZE, speakerPaint)
                        xOffset += speakerPaint.measureText(speakerText)
                    }

                    // Draw text with word wrapping
                    val availableWidth = PAGE_WIDTH - xOffset - MARGIN
                    val wrappedLines = wrapText(line.text, bodyPaint, availableWidth)

                    wrappedLines.forEachIndexed { index, wrappedLine ->
                        if (index > 0) {
                            yPosition += LINE_HEIGHT
                            canvas = ensureSpace(LINE_HEIGHT)
                            xOffset = MARGIN + 45f
                        }
                        canvas!!.drawText(wrappedLine, xOffset, yPosition + BODY_SIZE, bodyPaint)
                    }

                    yPosition += LINE_HEIGHT
                }
            }
            ContentType.TRANSLATION -> {
                translations?.forEach { segment ->
                    val range = if (segment.endLine != null && segment.endLine != segment.startLine) {
                        "[${segment.startLine}-${segment.endLine}]"
                    } else {
                        "[${segment.startLine}]"
                    }

                    // Calculate needed height for this segment
                    val fullText = if (!segment.speaker.isNullOrEmpty()) {
                        "$range [${segment.speaker}] ${segment.translationText}"
                    } else {
                        "$range ${segment.translationText}"
                    }

                    val availableWidth = PAGE_WIDTH - (MARGIN * 2)
                    val wrappedLines = wrapText(fullText, bodyPaint, availableWidth)

                    wrappedLines.forEachIndexed { index, wrappedLine ->
                        canvas = ensureSpace(LINE_HEIGHT)
                        canvas!!.drawText(wrappedLine, MARGIN, yPosition + BODY_SIZE, bodyPaint)
                        yPosition += LINE_HEIGHT
                    }

                    yPosition += LINE_HEIGHT * 0.5f // Extra space between segments
                }
            }
        }

        currentPage?.let { document.finishPage(it) }
        document.writeTo(outputStream)
        document.close()
    }

    private fun wrapText(text: String, paint: Paint, maxWidth: Float): List<String> {
        val words = text.split(" ")
        val lines = mutableListOf<String>()
        var currentLine = StringBuilder()

        for (word in words) {
            val testLine = if (currentLine.isEmpty()) word else "${currentLine} $word"
            val testWidth = paint.measureText(testLine)

            if (testWidth > maxWidth && currentLine.isNotEmpty()) {
                lines.add(currentLine.toString())
                currentLine = StringBuilder(word)
            } else {
                currentLine = StringBuilder(testLine)
            }
        }

        if (currentLine.isNotEmpty()) {
            lines.add(currentLine.toString())
        }

        return lines.ifEmpty { listOf("") }
    }
}

File: ios/ClassicsViewer/Utilities/TextExporter.swift
import Foundation
import UIKit
import PDFKit

enum ExportFormat {
    case csv
    case txt
    case pdf
}

enum ExportContentType {
    case source
    case translation
}

struct ExportRequest {
    let format: ExportFormat
    let contentType: ExportContentType
    let authorName: String
    let workTitle: String
    let bookLabel: String
    let workId: String
    let startLine: Int
    let endLine: Int
    let translator: String?
    let language: String
}

class TextExporter {

    // MARK: - CSV Export

    static func generateCsvContent(
        request: ExportRequest,
        lines: [TextLine]? = nil,
        translations: [TranslationSegment]? = nil
    ) -> String {
        var content = ""

        // CSV header
        content += "line_number,text\n"

        // Content
        switch request.contentType {
        case .source:
            if let lines = lines {
                for line in lines {
                    let text = csvEscape(line.lineText)
                    content += "\(line.lineNumber),\(text)\n"
                }
            }

        case .translation:
            if let translations = translations {
                for segment in translations {
                    let lineNum = segment.endLine != nil && segment.endLine != segment.startLine
                        ? "\(segment.startLine)-\(segment.endLine!)"
                        : "\(segment.startLine)"
                    let text = csvEscape(segment.translationText)
                    content += "\(lineNum),\(text)\n"
                }
            }
        }

        return content
    }

    private static func csvEscape(_ text: String) -> String {
        // If text contains comma, quote, or newline, wrap in quotes and escape quotes
        if text.contains(",") || text.contains("\"") || text.contains("\n") {
            let escaped = text.replacingOccurrences(of: "\"", with: "\"\"")
            return "\"\(escaped)\""
        }
        return text
    }

    // MARK: - TXT Export

    static func generateTxtContent(
        request: ExportRequest,
        lines: [TextLine]? = nil,
        translations: [TranslationSegment]? = nil
    ) -> String {
        var content = ""

        // Header
        content += "\(request.authorName) - \(request.workTitle)\n"
        content += "\(request.bookLabel), Lines \(request.startLine)-\(request.endLine)\n"
        if let translator = request.translator {
            content += "Translation by \(translator)\n"
        }
        content += "\n"

        // Content
        switch request.contentType {
        case .source:
            if let lines = lines {
                for line in lines {
                    var lineText = String(format: "%5d  ", line.lineNumber)
                    if let speaker = line.speaker, !speaker.isEmpty {
                        lineText += "[\(speaker)] "
                    }
                    lineText += line.lineText
                    content += lineText + "\n"
                }
            }

        case .translation:
            if let translations = translations {
                for segment in translations {
                    let range = segment.endLine != nil && segment.endLine != segment.startLine
                        ? "[\(segment.startLine)-\(segment.endLine!)]"
                        : "[\(segment.startLine)]"

                    var text = "\(range) "
                    if let speaker = segment.speaker, !speaker.isEmpty {
                        text += "[\(speaker)] "
                    }
                    text += segment.translationText
                    content += text + "\n\n"
                }
            }
        }

        return content
    }

    // MARK: - PDF Export

    static func generatePdf(
        request: ExportRequest,
        lines: [TextLine]? = nil,
        translations: [TranslationSegment]? = nil
    ) -> Data? {
        let pageWidth: CGFloat = 595  // A4
        let pageHeight: CGFloat = 842
        let margin: CGFloat = 50
        let contentWidth = pageWidth - (margin * 2)
        let lineHeight: CGFloat = 16

        let pdfMetaData = [
            kCGPDFContextCreator: "Classics Viewer" as CFString,
            kCGPDFContextAuthor: request.authorName as CFString,
            kCGPDFContextTitle: "\(request.workTitle) - \(request.bookLabel)" as CFString
        ]

        let format = UIGraphicsPDFRendererFormat()
        format.documentInfo = pdfMetaData as [String: Any]

        let renderer = UIGraphicsPDFRenderer(
            bounds: CGRect(x: 0, y: 0, width: pageWidth, height: pageHeight),
            format: format
        )

        // Use system fonts that support Greek/Latin
        let titleFont = UIFont.systemFont(ofSize: 18, weight: .bold)
        let subtitleFont = UIFont.systemFont(ofSize: 12)
        let bodyFont = UIFont.systemFont(ofSize: 11)

        let titleAttributes: [NSAttributedString.Key: Any] = [
            .font: titleFont,
            .foregroundColor: UIColor.black
        ]

        let subtitleAttributes: [NSAttributedString.Key: Any] = [
            .font: subtitleFont,
            .foregroundColor: UIColor.darkGray
        ]

        let bodyAttributes: [NSAttributedString.Key: Any] = [
            .font: bodyFont,
            .foregroundColor: UIColor.black
        ]

        let data = renderer.pdfData { context in
            var yPosition: CGFloat = margin

            func startNewPage() {
                context.beginPage()
                yPosition = margin
            }

            func ensureSpace(_ needed: CGFloat) {
                if yPosition + needed > pageHeight - margin {
                    startNewPage()
                }
            }

            func drawText(_ text: String, attributes: [NSAttributedString.Key: Any], at point: CGPoint, maxWidth: CGFloat) -> CGFloat {
                let attributedString = NSAttributedString(string: text, attributes: attributes)
                let boundingRect = attributedString.boundingRect(
                    with: CGSize(width: maxWidth, height: .greatestFiniteMagnitude),
                    options: [.usesLineFragmentOrigin, .usesFontLeading],
                    context: nil
                )
                ensureSpace(boundingRect.height)
                attributedString.draw(in: CGRect(x: point.x, y: yPosition, width: maxWidth, height: boundingRect.height))
                return boundingRect.height
            }

            startNewPage()

            // Title
            yPosition += drawText(
                "\(request.authorName) - \(request.workTitle)",
                attributes: titleAttributes,
                at: CGPoint(x: margin, y: yPosition),
                maxWidth: contentWidth
            )
            yPosition += 8

            // Subtitle
            let subtitle = "\(request.bookLabel), Lines \(request.startLine)-\(request.endLine)"
            yPosition += drawText(subtitle, attributes: subtitleAttributes, at: CGPoint(x: margin, y: yPosition), maxWidth: contentWidth)
            yPosition += 4

            if let translator = request.translator {
                yPosition += drawText("Translation by \(translator)", attributes: subtitleAttributes, at: CGPoint(x: margin, y: yPosition), maxWidth: contentWidth)
                yPosition += 4
            }

            yPosition += drawText("Exported from \(request.workId)", attributes: subtitleAttributes, at: CGPoint(x: margin, y: yPosition), maxWidth: contentWidth)
            yPosition += 24

            // Content
            switch request.contentType {
            case .source:
                lines?.forEach { line in
                    let lineNum = String(format: "%5d", line.lineNumber)
                    var text = "\(lineNum)  "
                    if let speaker = line.speaker, !speaker.isEmpty {
                        text += "[\(speaker)] "
                    }
                    text += line.lineText

                    yPosition += drawText(text, attributes: bodyAttributes, at: CGPoint(x: margin, y: yPosition), maxWidth: contentWidth)
                    yPosition += 4
                }

            case .translation:
                translations?.forEach { segment in
                    let range = segment.endLine != nil && segment.endLine != segment.startLine
                        ? "[\(segment.startLine)-\(segment.endLine!)]"
                        : "[\(segment.startLine)]"

                    var text = "\(range) "
                    if let speaker = segment.speaker, !speaker.isEmpty {
                        text += "[\(speaker)] "
                    }
                    text += segment.translationText

                    yPosition += drawText(text, attributes: bodyAttributes, at: CGPoint(x: margin, y: yPosition), maxWidth: contentWidth)
                    yPosition += lineHeight
                }
            }
        }

        return data
    }
}
